Provides extension health via builditem

loicmathieu opened the pull request:
This PR implements health check inside the Agroal extension so a user will have an health check that validate the database if it includes the agroal extension.

I create a `HealthBuildItem` that an extension needs to emmit to inlcude it's health check. Smallrye Heath extension will use these `HealthBuildItem`s to add the health check to ArC if extension health check is enabled (true by default).

It can be disabled with the following properties:
```
# disable the datasource health check
quarkus.datasource.health.enabled=false
# disable all extension health check
quarkus.health.extensions.enabled=false
```

This is part of #3107

Ping @emmanuelbernard @kenfinnigan @stuartwdouglas

loicmathieu committed:
feat: add an healthcheck to agroal

loicmathieu committed:
feat: extensions health check

machi1990 commented on extensions/agroal/deployment/pom.xml:
```diff
@@ -31,6 +31,13 @@
+        <!-- Add the health extension as optional as we will produce the health check only if it's included -->
+        <dependency>
+            <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-smallrye-health</artifactId>
```
is `-deployment` missing from the artifactId?

loicmathieu replied:
Agroal only needs the runtime dependency to be able to compile the health check. Moreover the dependency is optional to avoid adding it in case agroal is used without smallrye-health.

In fact we can witch to io.smallrye:smallrye-health dependency instead of Quarkus one if we want as there is no dependency to the extention ...

machi1990 commented on extensions/agroal/runtime/src/main/java/io/quarkus/agroal/runtime/health/DataSourceHealthCheck.java:
```diff
@@ -0,0 +1,36 @@
+            try {
+                boolean valid = dataSource.getConnection().isValid(0);
+                if (!valid) {
+                    builder.down().withData("reason", "validation check failed");
```
I think this will be overridden if two datasources are down? 
Since we support multiple datasources, will it be possible to return an object having this structure in case of failures? 
``` 
{
 "reason": {
    "validation check failed",
     "datasources": ["data-source-name-1", "data-source-name-2", ..., "data-source-name-N"]
  }
} 
```

loicmathieu replied:
In case multiple datasources down, there will be multiple data with the same content as the `withData` method adds the data to a HashMap.

First I want to implement exactly what you ask ... then I realize that I don't know how to retrieve the beans with there name in Arc :) 

In case multiple beans of the same type I assume they will have different qualifier. But in CDI to retrieve multiple beans of the same type you use `Instance<Datasource>` that provides the list of the beans and not there names ... If someone can explain me how to retrieve the list of beans from ArC with there qualifier (name) I'll update my implementation.

loicmathieu replied:
Totaly agree but as I reply to @machi1990 just above, I don't know how to retrieve the bean names from ArC, I use `@Inject Instance<DataSource> dataSources;` that gives me all datasources but not the bean name/qualifier ...

machi1990 replied:
> ... If someone can explain me how to retrieve the list of beans from ArC with there qualifier (name) I'll update my implementation.

@loicmathieu I am certainly not a CDI expert, but have you tried something like this? 
```
 final Set<Bean<?>> beans = Arc.container().beanManager().getBeans(javax.sql.DataSource.class);
        for (Bean<?> bean : beans) {
            if (bean.getName() == null) {
                //  default DataSource has not name can be retrieved via
                //  @Inject
                //  javax.sql.DataSource defaultDataSource;

                //TODO do something with defaultDataSource
            } else {
                InstanceHandle<javax.sql.DataSource> instanceHandle = Arc.container().instance(bean.getName());
                if (instanceHandle.isAvailable()) {
                    javax.sql.DataSource dataSource = instanceHandle.get();
                    // TODO do something with dataSource
                }
            }
        }
```
Also pinging @mkouba @manovotn, may be there is a better way to achieve this  :-)

manovotn replied:
Your solution is viable, otherwise you could use `@Any Instance<DataSource>` which would bring you all beans of type `DataSource` regardless of their qualifiers. You can then iterate over all of them and inspect them.

If you are after bean names (and not `DataSource` instances), then going via `Arc.container().beanManager().getBeans(...)` is better.

loicmathieu replied:
Thanks @machi1990 and @manovotn I used `Arc.container().beanManager().getBeans(...)` in a `  @PostConstruct` and it works fine!

loicmathieu replied:
@kenfinnigan the first argument of `withData` should be a valid json attribute, for the moment I ended up generating `reason1`, `reason2`, ... but maybe `datasourceName-check-status` will be better but the default datasource has no name and if I put `default` it can collide with someone that use a named datasource of name `default`...

loicmathieu replied:
No, it surprises me so I check but you can have a default datasource (with no name) and a nammed datasource. See [Multiple Datasource](https://quarkus.io/guides/datasource-guide#multiple-datasources).

So I can configure something like this:
```
# default un-named datasource : can be injected with @Inject with no other annotation
quarkus.datasource.driver=org.h2.Driver
quarkus.datasource.url=jdbc:h2:tcp://localhost/mem:default
quarkus.datasource.username=username-default
quarkus.datasource.min-size=3
quarkus.datasource.max-size=13

# Named datasource with the name default, must be injected with @Datasource annotation
quarkus.datasource.default.driver=org.h2.Driver
quarkus.datasource.default.url=jdbc:h2:tcp://localhost/mem:users
quarkus.datasource.default.username=username1
quarkus.datasource.default.min-size=1
quarkus.datasource.default.max-size=11
```

loicmathieu replied:
It's just a bit misleading for the developer to talk about the datasource named default in this case: do it refers to the default unamed datasource or the named default datasource ? My proposal clearly talk about the default datasource or the datasource named 'default' in this case.

loicmathieu replied:
You can, the default datasource has no name (name = null)

machi1990 commented on extensions/agroal/runtime/src/main/java/io/quarkus/agroal/runtime/AgroalBuildTimeConfig.java:
```diff
@@ -20,4 +20,10 @@
+    /**
+     * Whether or not an healtcheck is published in case the smallrye-health extionsion is present (default to true).
+     */
+    @ConfigItem(name = "health.enabled", defaultValue = "true")
```
I wonder if leaving out this naming the config item and having a default name `health-enabled` would have been a viable alternative? 

Also would one of `enable-healthcheck` or `healthcheck-enabled` would be better name for this config item?

Should the default value be controlled by whatever is set in `HealthBuildTimeConfig#extensionsEnabled`?

loicmathieu replied:
This is a design choice still opens to discussion ;) I know that @emmanuelbernard want that as much as possible works will be included inside the smallrye-health extension but I think that it's the agroal extension that needs to control it's configuration and the integration of the health check: this will be less works to do during the augmenation phase ...

Your proposal to add `HealthBuildTimeConfig#extensionsEnabled` will not facilitate the usage for the agroal extension a lot as it will still need to support the configuration item ...

machi1990 commented on extensions/smallrye-health/runtime/src/main/java/io/quarkus/smallrye/health/runtime/HealthBuildTimeConfig.java:
```diff
@@ -0,0 +1,15 @@
+     */
+    // This config item is used by extensions to avoid to publish the health check but not directly used in the health
+    // extension. It must exists to avoid a log WARN of 'Unrecognized configuration key'.
+    @ConfigItem(name = "extensions.enabled", defaultValue = "true")
```
How does renaming this to `enableExtensionCheck` with the config key being `enable-extension-check` sound? 

Will defaulting to hyphenated config key name  be a viable naming? e.g taking for example the present variable naming i.e `extensionsEnabled ` the configuration key will be `extensions-enabled`?

loicmathieu replied:
These are the configuration names proposed in the original issue ... I'm open to suggestion but I found them easy to use and honestly I'm not a big fan of hyphenated names ...

machi1990 reviewed (changes requested):
@loicmathieu I just have a couple of questions and suggestions. Great PR otherwise.

kenfinnigan reviewed (changes requested):
Would be good if @xstefank and @antoinesd also looked at this

loicmathieu commented on extensions/smallrye-health/runtime/src/main/java/io/quarkus/smallrye/health/runtime/HealthBuildTimeConfig.java:
```diff
@@ -0,0 +1,15 @@
+import io.quarkus.runtime.annotations.ConfigRoot;
+
+@ConfigRoot(name = "health")
+public class HealthBuildTimeConfig {
```
It confuses me also but I copied it from other extensions that separate runtime and buildtime config in two classes, both in the runtime artifact.

IMHO, `SmallRyeHealthProcessor.SmallRyeHealthConfig` is smallrye related (config root is `quarkus.smallrye-health`) and here it's Quarkus specific stuff (config root is `quarkus.health`).

dmlloyd replied:
The build time config should normally be in the `deployment` artifact.  Here's the breakdown:

* `BUILD_TIME` (the default) - `deployment` artifact (or possibly `spi` in some future cases)
* `BUILD_AND_RUN_TIME_FIXED` - `runtime` artifact always
* `RUN_TIME` - `runtime` artifact always

dmlloyd replied:
Also it's 100% OK to have more than one config root use the same name and overlap arbitrarily, whether in the same or in different config phases, as long as no *leaf* configuration values end up with conflicting names.

loicmathieu replied:
@kenfinnigan, @dmlloyd will move it to the deployment artifact next week (I'm leaving now : week-end :) ).
As for the name, it's part of a global naming discussed during the design of the feature, I'll wait for @emmanuelbernard review on the design.

loicmathieu replied:
@kenfinnigan @dmlloyd the build time config has been moved to the deployment module.

loicmathieu commented on extensions/agroal/runtime/src/main/java/io/quarkus/agroal/runtime/health/DataSourceHealthCheck.java:
```diff
@@ -0,0 +1,36 @@
+                    builder.down().withData("reason", "validation check failed");
+                }
+            } catch (SQLException e) {
+                builder.down().withData("reason", "SQLException trying to validate the connection: " + e.getMessage());
```
Maybe `"Unable to execute the health check: " + e.getMessage` ?

loicmathieu commented:
@kenfinnigan @machi1990 I add some documentation.
I didn't add documentation to the extension guide as there is no extension specific section in the extension guide but maybe it worth it ?

machi1990 commented on docs/src/main/asciidoc/health-guide.adoc:
```diff
@@ -296,6 +296,15 @@ error along with the health check response.
+
+Some extension may provides default health checks, including the extension will automatically register its health checks.
+
+For example, `quarkus-agroal` that is use to manage Quarkus datasource(s) automatically register a readiness health check
```
There is a typo .. `s/that is use/that is used`

loicmathieu commented:
@kenfinnigan I add a section in the extension guide. 
I found this guide very complex and I have a hard time decided where to include this new section so I'm not sure it was the best place to put it ...

loicmathieu commented:
@kenfinnigan I didn't provide any tests, should I create an health check on the test-extension and validate of the test-extension integration test that the health check is successfully integrated ?

dmlloyd commented:
Let's follow the general approach of putting shared build items into an extension-specific SPI module.  This way the build items can be shared without introducing hard dependencies on the extension deployment modules.

loicmathieu commented:
@dmlloyd I implemented the SPI approach, can you check if it's correclty implemented ?

kenfinnigan commented on docs/src/main/asciidoc/datasource-guide.adoc:
```diff
@@ -223,6 +223,16 @@ AgroalDataSource dataSource1;
+to validate the datasource.
+
+So when you access the `/health/ready` endpoint of your application you will have information about the datasource validation status.
+If you have multiple datasources, all datasources will be checked and the status will be `DOWN` as soon as there is one datasource validation failed.
```
Change "validation failed" to "validation failure"

loicmathieu replied:
done

kenfinnigan commented on docs/src/main/asciidoc/extension-authors-guide.adoc:
```diff
@@ -874,6 +874,37 @@ As a CDI based runtime, Quarkus extensions often make CDI beans available as par
 
+=== Extension Health Check
+
+Health check is provided via the `quarkus-smallrye-health` extension. It provides both liveness and readiness checks capabilities.
```
"Health check is" to "Health checks are"

loicmathieu replied:
done

kenfinnigan commented on docs/src/main/asciidoc/extension-authors-guide.adoc:
```diff
@@ -874,6 +874,37 @@ As a CDI based runtime, Quarkus extensions often make CDI beans available as par
+
+Health check is provided via the `quarkus-smallrye-health` extension. It provides both liveness and readiness checks capabilities.
+
+When you write an extension, it can be very handy to provides a default health check for your extension to prevent the developer to do it
```
Maybe reword to "When writing an extension, it's beneficial to provide health checks for the extension that can be automatically included without the developer needing to write their own". Or something like that

loicmathieu replied:
done

kenfinnigan commented on docs/src/main/asciidoc/extension-authors-guide.adoc:
```diff
@@ -874,6 +874,37 @@ As a CDI based runtime, Quarkus extensions often make CDI beans available as par
+In order to provide a health check, you should do the following:
+
+- Import the `quarkus-smallrye-health` extension as an **optional** dependency in your runtime module so it will not impact the size of the application if
+health check is not needed.
```
Should that be "not needed" or "not included"?

loicmathieu replied:
done

kenfinnigan commented on docs/src/main/asciidoc/health-guide.adoc:
```diff
@@ -296,6 +296,15 @@ error along with the health check response.
 
+== Extension health checks
+
+Some extension may provides default health checks, including the extension will automatically register its health checks.
```
Change "extension may provides" to "extensions may provide".

loicmathieu replied:
done

kenfinnigan commented on extensions/agroal/runtime/src/main/java/io/quarkus/agroal/runtime/health/DataSourceHealthCheck.java:
```diff
@@ -0,0 +1,62 @@
+                String data = dataSource.getKey() == DEFAULT_DS
+                        ? "Unable to execute the validation check for the default DataSource: "
+                        : "Unable to execute the validation check for DataSource '" + dataSource.getKey() + "': ";
+                builder.down().withData("reason" + cpt++, data + e.getMessage());
```
I still think we should use a key other than "reason" to indicate an exception occurred.

loicmathieu replied:
OK, I propose to use the name of the datasource or 'quarkus-default-ds' for the default one. 
It's very unlinkly that a user will create a named datasource with 'quarkus-default-ds' as its name.

dmlloyd reviewed (commented):
Seems OK to me: in terms of the SPI structure that is.  I'll let the other reviews stand.

dmlloyd commented on extensions/agroal/runtime/src/main/java/io/quarkus/agroal/runtime/health/DataSourceHealthCheck.java:
```diff
@@ -0,0 +1,62 @@
+            try {
+                boolean valid = dataSource.getValue().getConnection().isValid(0);
+                if (!valid) {
+                    String data = dataSource.getKey() == DEFAULT_DS ? "validation check failed for the default DataSource"
```
String compare should always use `equals`.

loicmathieu replied:
Arrrhgggg such a newbie error! I'm really ashame as I think I didn't do this error since at least 10 yers !

loicmathieu replied:
fixed

loicmathieu replied:
For the record, it worked because it's the same string constant that is defined on the class and put as hashmap key so it's valid to use `==` but I'm agree that this should not be used!

dmlloyd commented on extensions/agroal/runtime/src/main/java/io/quarkus/agroal/runtime/health/DataSourceHealthCheck.java:
```diff
@@ -0,0 +1,62 @@
+                    builder.down().withData("reason" + cpt++, data);
+                }
+            } catch (SQLException e) {
+                String data = dataSource.getKey() == DEFAULT_DS
```
Another `equals`.

loicmathieu replied:
Yep, I saw all of them :=

loicmathieu replied:
fixed

machi1990 commented on extensions/agroal/runtime/src/main/java/io/quarkus/agroal/runtime/health/DataSourceHealthCheck.java:
```diff
@@ -0,0 +1,63 @@
+            try {
+                boolean valid = dataSource.getValue().getConnection().isValid(0);
+                if (!valid) {
+                    String data = DEFAULT_DS.equals(dataSource.getKey())  ? "validation check failed for the default DataSource"
```
Maybe we can declare a variable for `DEFAULT_DS.equals(dataSource.getKey())`, seems to be done on multiple places.

loicmathieu replied:
done

loicmathieu committed:
doc: guides

loicmathieu commented:
@kenfinnigan is the last batch of changes OK for you ? Can you advise on what should be done for testing it ? (adding an health check in the test-extension extension ?)

@emmanuelbernard can you validate the design regarding what was discussed for the issue #3107

emmanuelbernard commented:
I’m swamped man. So I can trust your work or have somebody quickly look it
up. If you have the doc side of this PR i could definitely read it quick
and see if any alarm bell rings :)

On Tue 3 Sep 2019 at 16:53, Loïc Mathieu <notifications@github.com> wrote:

> @kenfinnigan <https://github.com/kenfinnigan> is the last batch of
> changes OK for you ? Can you advise on what should be done for testing it ?
> (adding an health check in the test-extension extension ?)
>
> @emmanuelbernard <https://github.com/emmanuelbernard> can you validate
> the design regarding what was discussed for the issue #3107
> <https://github.com/quarkusio/quarkus/issues/3107>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/quarkusio/quarkus/pull/3671?email_source=notifications&email_token=AACJNWFMSFOLEZ2HDXVOFIDQHZ265A5CNFSM4IO7KTCKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD5YO4QY#issuecomment-527494723>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AACJNWARY4ICIZXOMGJCRATQHZ265ANCNFSM4IO7KTCA>
> .
>

kenfinnigan commented:
@loicmathieu I think it looks good. In terms of a test, I *think* there's an integration-test that uses a database so we could have a test that ensures the Agroal health check is present?

loicmathieu commented:
@emmanuelbernard I understand that :) that's why I ask to check the design, @kenfinnigan already works with me on the implementation side! 

You can have a look at the following documentation parts:
- Datasource guide: https://github.com/quarkusio/quarkus/pull/3671/files#diff-eb196d770f1a1f55ee5d373bdd05646a
- Health guide: https://github.com/quarkusio/quarkus/pull/3671/files#diff-4e46b3e4681926690297d644df6344fb
- Extension author guide: https://github.com/quarkusio/quarkus/pull/3671/files#diff-38cd571a6836972129eb3e2eb8528382

 @kenfinnigan the main integration test has both datasource and health extension so it will be as easy as adding an assert to it ;). I also integrate an health check to the test-extension extesion, it's up to you wether I push it or not !

loicmathieu commented:
@kenfinnigan I add a test on the main integration test. I also rebase on master.

FroMage commented:
@kenfinnigan ball is in your camp :)

stuartwdouglas commented on extensions/agroal/runtime/src/main/java/io/quarkus/agroal/runtime/health/DataSourceHealthCheck.java:
```diff
@@ -0,0 +1,63 @@
+        for (Map.Entry<String, DataSource> dataSource : dataSources.entrySet()) {
+            boolean isDefault = DEFAULT_DS.equals(dataSource.getKey());
+            try {
+                boolean valid = dataSource.getValue().getConnection().isValid(0);
```
Don't we need to close the connection here to release it back to the pool?

loicmathieu replied:
Yes, fixed

loicmathieu replied:
@stuartwdouglas is the fix OK for you ?

emmanuelbernard commented on docs/src/main/asciidoc/extension-authors-guide.adoc:
```diff
@@ -874,6 +874,36 @@ As a CDI based runtime, Quarkus extensions often make CDI beans available as par
+----
+@BuildStep
+HealthBuildItem addHealthCheck(AgroalBuildTimeConfig agroalBuildTimeConfig) {
+    if (agroalBuildTimeConfig.healthEnabled) {
```
My main point is that all extensions would need to do that : read the flag and do the if. And when we introduce filtering with wildcard and all, it will be even more conplicated.

loicmathieu replied:
I add a boolean on the `HealthBuildItem` so that each extension will only needs to pass it's configuration item to it.

At the begining I wanted to retrieve the configuration item in smallrye-health via `ConfigProvider.getConfig()` but it didn't works, I was not able to retrieve the configuration item during the build step ...

Another solution will be to change the way all this is configured from :
```
quarkus.health.extensions.enabled=true # provided by smallrye-health
quarkus.datasource.health.enabled=false # provided by **agroal**
```

To
```
quarkus.health.extensions.enabled=true # provided by smallrye-health
quarkus.health.extensions.datasource.enabled=false # provided by **smallrye-health**
```

emmanuelbernard replied:
> I add a boolean on the HealthBuildItem so that each extension will only needs to pass it's configuration item to it.

What do you mean by that? You lost me I think.

emmanuelbernard replied:
Ah I think I see now, you mean this https://github.com/quarkusio/quarkus/pull/3671/files#diff-38cd571a6836972129eb3e2eb8528382R901

emmanuelbernard replied:
Sounds ok, if reading properties from the SR extension is not an option.

emmanuelbernard commented on extensions/smallrye-health/spi/src/main/java/io/quarkus/smallrye/health/deployment/spi/HealthBuildItem.java:
```diff
@@ -0,0 +1,15 @@
+public final class HealthBuildItem extends MultiBuildItem {
+    private Class<?> healthCheckClass;
+
+    public HealthBuildItem(Class<?> healthCheckClass) {
```
I think a health builditem could have a name: e.g. `quarkus.agroal.database_connection` (or something else). this is how we could do selective filtering down the line

loicmathieu replied:
As the name is not currently used I didn't incude it already so we will nee to ask for it when filtering will be implemented.

emmanuelbernard replied:
I was about to have it now as this will be a breaking change for extensions and wanted to avoid that but if we are unstable on the name concept, that's premature.

loicmathieu replied:
I was asking myself the same: why not to add it now. 
But I don't think breaking change on extension is really an issue as we can revisit all extension that uses it at the same time. It's an internal breaking change that didn't impact users so it's OK.

emmanuelbernard reviewed (changes requested):
Looks good except my remark on reuse of the filtering code

loicmathieu committed:
fix:  avoid testing the inclusion in each extension

Also some minor doc fixes

loicmathieu commented:
@emmanuelbernard is it OK for you ?

machi1990 commented:
@loicmathieu just a minor comment, can we squash the commits? thanks.

loicmathieu commented:
@machi1990 I prefere to wait when all is OK before squashing the commits. Still waiting for some feedback ...

loicmathieu commented:
@machi1990 I squashed some commits but I keep the last one as there was too many conflit.
@kenfinnigan I think that all points are covered by the current implementation so it is good to go.

kenfinnigan commented:
I think we need @emmanuelbernard and @stuartwdouglas to review the changes they requested

emmanuelbernard commented:
On 13 Sep 2019, at 15:56, Loïc Mathieu wrote:

> loicmathieu commented on this pull request.
>
>
>
>> @@ -0,0 +1,15 @@
> +package io.quarkus.smallrye.health.deployment.spi;
> +
> +import io.quarkus.builder.item.MultiBuildItem;
> +
> +public final class HealthBuildItem extends MultiBuildItem {
> +    private Class<?> healthCheckClass;
> +
> +    public HealthBuildItem(Class<?> healthCheckClass) {
>
> I was asking myself the same: why not to add it now.
> But I don't think breaking change on extension is really an issue as 
> we can revisit all extension that uses it at the same time. It's an 
> internal breaking change that didn't impact users so it's OK.
>

That is not true, we have external extensions, Camel moved out of the 
Quarkus repo for example and that trend will only accelerate.

loicmathieu commented:
I didn't thought about external extension.  So do you want to add the name now to avoid a future breaking change later ?

loicmathieu commented:
> I didn't thought about external extension. So do you want to add the name now to avoid a future breaking change later ?

@emmanuelbernard do you want me to add the name of the extension in the `HealthBuildItem for future use?

loicmathieu committed:
feat: add configRootName

Also switch to class name instead of class object, it is the usage on Quarkus deployment module and can avoid unecessary class loading.

loicmathieu committed:
doc: update the doc to reflect the last changes

loicmathieu commented:
@kenfinnigan I discuss with @emmanuelbernard on zulip and we agree to add the `configRootName` for future usage.

I also switch to use class name instead of class object because it is the prefered usage across Quarkus extension and, as I understand it, avoid possible unecessary class loading.

I also rebase onto master.

I think that everything is OK so for me it's ready to be merged. If we agree on this I will squash on two commits (one for agroal health check and one for smallrye-heath).
